gateway/src/apicast/policy/default_credentials/default_credentials.lua

```diff
@@ -0,0 +1,98 @@
+--- Default credentials policy
+
+local tostring = tostring
+
+local policy = require('apicast.policy')
+local _M = policy.new('Default credentials policy')
+
+local new = _M.new
+
+function _M.new(config)
+  local self = new(config)
+
+  if config then
+    self.default_credentials = {
+      user_key = config.user_key,
+      app_id = config.app_id,
+      app_key = config.app_key
+    }
+  else
+    self.default_credentials = {}
+  end
+
+  return self
+end
+
+local function creds_missing(service)
+  local service_creds = service:extract_credentials()
+  if not service_creds then return true end
+
+  local backend_version = tostring(service.backend_version)
+
+  if backend_version == '1' then
+    return service_creds.user_key == nil
+  elseif backend_version == '2' then
+    return service_creds.app_id == nil and service_creds.app_key == nil
+  end
+end
+
+local function provide_creds_for_version_1(service, default_creds)
+  if default_creds.user_key then
+    -- follows same format as Service.extract_credentials()
+    service.extracted_credentials = {
+      default_creds.user_key,
+      user_key = default_creds.user_key
+    }
+
+    ngx.log(ngx.DEBUG, 'Provided default creds for request')
+  else
+    ngx.log(ngx.WARN, 'No default user key configured')
+  end
+end
+
+local function provide_creds_for_version_2(service, default_creds)
+  if default_creds.app_id and default_creds.app_key then
+    -- follows same format as Service.extract_credentials()
+    service.extracted_credentials = {
+      default_creds.app_id,
+      default_creds.app_key,
+      app_id = default_creds.app_id,
+      app_key = default_creds.app_key
+    }
+
+    ngx.log(ngx.DEBUG, 'Provided default creds for request')
+  else
+    ngx.log(ngx.WARN, 'No default app_id + app_key configured')
+  end
+end
+
+local creds_provider = {
+  ["1"] = provide_creds_for_version_1,
+  ["2"] = provide_creds_for_version_2
+}
+
+local function provide_creds(service, default_creds)
+  local backend_version = tostring(service.backend_version)
+  creds_provider[backend_version](service, default_creds)
+end
+
+function _M:rewrite(context)
+  local service = context.service
+
+  if not service then
+    ngx.log(ngx.ERR, 'No service in the context')
+    return
+  end
+
+  local backend_version = tostring(service.backend_version)
+  if backend_version ~= '1' and backend_version ~= '2' then
```

Comment:
This duplicates the `creds_provider`. Imo it should check if there is creds provider and then run it else return an error instead of duplicating the checks.

Reply:
But the method that gets called before `provide_creds` (`missing_creds`), also needs to check if the `backend_version` is supported by the policy.

Doing the check where it's done now allows us to assume in the rest of the code (`missing_creds` and `provide_creds`) that the backend version is supported. If we don't do that, we'll need to add checks and `ngx.log(ngx.ERR, ...`) in those 2 methods.

Reply:
But it still can be changed to `if not creds_provider[backend_version]`, right?

My point is that if you remove/add element to `creds_provider` this line would also have to change.

Reply:
OK. I rewrote this in a way that adding/removing supported backend_versions would imply changes only in one part of the code.